File: Haskell/problem010.hs
{-
Problem 10 - Project Euler
http://projecteuler.net/index.php?section=problems&id=10
-}

sieve :: Int -> [Int] -> [Int] -> [Int]
sieve _ _ [] = []
sieve m xs (y : ys)
  | y * y >= m = reverse xs ++ y : ys
  | otherwise = sieve m (y : xs) (filter (\a -> a `mod` y /= 0) ys)

primesBelow x = sieve x [] [2, 3 .. x]

problem010 = sum . primesBelow $ (2 * 1000 * 1000)

main = putStrLn $ "problem010: answer " ++ show problem010

File: Elixir/problem010.exs
# Problem 10 - Project Euler
# http://projecteuler.net/index.php?section=problems&id=10
defmodule P010 do
  defp sieve([hd | tl], x, ans) do
    if hd * hd > x do
      Enum.reverse(ans) ++ [hd | tl]
    else
      tl
      |> Enum.filter(&(rem(&1, hd) != 0))
      |> sieve(x, [hd | ans])
    end
  end

  def solve() do
    x = 2_000_000
    2..x - 1
    |> Enum.into([])
    |> sieve(x, [])
    |> Enum.sum
  end
end

IO.puts P010.solve()
